File: 2015/1/Go/main.go
package main

import (
	"errors"
	"fmt"
	"os"
)

func calcFloor(content string) int {
	floor := 0
	// Standard loop wont work unless converted to a rune array, as strings are byte sequences,
	// thing like "a£c" contains 4 bytes not 3
	// A for-range loop will work too
	runeContent := []rune(content)
	for i := 0; i < len(runeContent); i++ {
		if content[i] == '(' {
			floor++
		} else {
			floor--
		}
	}
	return floor
}

func calcBasementEnteringPos(content string) (*int, error) {
	floor := 0
	var pos *int
	var val int

	for idx, char := range content {
		if char == '(' {
			floor++
		} else if char == ')' {
			floor--
		}
		if floor == -1 {
			val = idx + 1
			pos = &val
			return pos, nil
		}
	}
	return nil, errors.New("unable to enter basement")
}

func readFileContents(filepath string) (string, error) {
	content, err := os.ReadFile(filepath)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func main() {
	filepath := "input.txt"
	content, err := readFileContents(filepath)
	if err != nil {
		fmt.Printf("Unable to read the file %s: %s\n", filepath, err)
		return
	}
	// First challenge
	floor := calcFloor(content)
	fmt.Printf("The floor Stanta needs to reach is %d\n", floor)
	// Second challenge
	basementPos, err := calcBasementEnteringPos(content)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Printf("The position of the character to enter the basement is: %d\n", *basementPos)

	}
}

File: 2015/1/Zig/src/main.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const GeneralPurposeAllocator = std.heap.GeneralPurposeAllocator;

pub fn calcFloor(content: []const u8) isize {
    var floor: isize = 0;
    for (content) |char| {
        if (char == '(') {
            floor += 1;
        } else if (char == ')') {
            floor -= 1;
        }
    }
    return floor;
}

pub fn calcBasementEnteringPos(content: []const u8) ?usize {
    var floor: isize = 0;
    var pos: ?usize = null;
    for (content, 0..) |char, index| {
        if (char == '(') {
            floor += 1;
        } else if (char == ')') {
            floor -= 1;
        }
        if (floor == -1) {
            pos = index + 1; // Position start at index 1
            break;
        }
    }
    return pos;
}

pub fn readFileContents(allocator: *std.mem.Allocator, input_path: []const u8) ![]u8 {
    // Open the file for reading
    const file = try std.fs.cwd().openFile(input_path, .{ .mode = .read_only });
    defer file.close();

    // Determine the size of the file
    const fileSize = try file.getEndPos();

    // Allocate a buffer to hold the file contents
    var buffer = try allocator.alloc(u8, fileSize);

    // Read the file into the buffer
    _ = try file.readAll(buffer);

    // Return a slice of the buffer containing the read data
    return buffer[0..fileSize];
}

pub fn main() !void {
    const filepath = "input.txt";

    var gpa = GeneralPurposeAllocator(.{}){}; // Instantiate the allocator
    var allocator = gpa.allocator(); // Obtain the allocator interface as a mutable variable
    defer {
        const deinitStatus = gpa.deinit();
        if (deinitStatus == .leak) {
            std.debug.print("Memory leak detected\n", .{});
        }
    }
    const content_buffer = try readFileContents(&allocator, filepath);
    defer allocator.free(content_buffer);

    const floor = calcFloor(content_buffer);
    std.debug.print("The floor Santa needs to reach is: {}\n", .{floor});

    const basement_pos: ?usize = calcBasementEnteringPos(content_buffer);
    std.debug.print("The position of the character to enter the basement is: {?}\n", .{basement_pos});
}

test "first challenge" {
    const test_content = "))(((((";
    const floor = calcFloor(test_content);
    std.debug.print("The floor Santa needs to reach is: {}\n", .{floor});
}

test "second challenge" {
    const test_content = "()())";
    const basement_pos: ?usize = calcBasementEnteringPos(test_content);
    std.debug.print("The position of the character to enter the basement is: {?}\n", .{basement_pos});
}
